### ISSUE TITLE
Split both options of no-unused-modules into 2 rules

### ISSUE THREAD

**Issue by astorije** [benmosher/eslint-plugin-import#1684]
Apologies if there is already a ticket for this.

I find `no-unused-modules` a very powerful rule however very confusing due to its options. It seems to me `missingExports` and `unusedExports` effectively respond to 2 different use cases.

Additionally, the `ignoreExports` option applies to both options, but sometimes it makes sense to ignore patterns for one but not for the other.
For example, it might be ok for everything under `test/` to not export anything but we still might want to make sure that anything exported in a test file is imported somewhere (another test file for example). For this use case, my organization uses a dedicated `.eslintrc.yml` file in `test/`, but people locating their test files next to their source files will not have this chance.

I believe I saw a ticket suggesting an additional ignore list to address this, which would help making my case, but I can't seem to find it anymore so maybe I just saw what I wanted to see 😅

Lastly, I find this rule difficult to configure and predict: it respects `.eslintignore` patters, nested ESLint configurations, has an additional `src` options to find further patterns, but also an ignore list on its own, and will even honor [files listed in `package.json`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unused-modules.md#important-note) (which is potentially great) _if and only if_ the project is not private (which is less great).
I feel like splitting this rule in 2 would help reduce the confusion and may offer a way to simplify branches of configuration/ignores.

Thoughts?

**Comment by astorije** [comment]
Oh and I forgot, the rule is present in both the [Static analysis](https://github.com/benmosher/eslint-plugin-import#static-analysis) and [helpful warning](https://github.com/benmosher/eslint-plugin-import#helpful-warnings) of the README, which I'm sure is an accident, but thought it was a funny coincidence 😅

**Comment by ljharb** [comment]
The primary use cases are for an app (private: true) vs a library (has a "main"/"exports").

Both need a way to ensure tests/examples/storybook stories/etc don't count as usage (a rule-specific configuration).

Both need an option to determine if you want to check for individual export usage vs "anything in the file" usage.

For a library, the rule needs a way to indicate "entry points" - things that are ok to be unused, because users might be using them, so it would be a breaking change if someone deleted it because of this eslint rule.

This rule - like a number of the ones in this plugin - are not really practical to use in a shared config, because they are so tightly coupled to the specific repo they're used in.

I'm not sure how much clarity it would provide to have an "no unused modules in an app" rule vs a "no unused modules in a library" rule, and they'd have heavy overlap.

**Comment by astorije** [comment]
I'll assume I understood your comment in my response, so apologies if I write something that shows I didn't 😅

The split I was thinking of was not really about apps vs. libs, but rather separating the [2 points in the intro of the rule](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unused-modules.md#importno-unused-modules): one rule that checks for files without exports, and one rule that checks if exports are used somewhere in the project.

That way both could be configured differently based on paths/globs.

**Comment by ljharb** [comment]
Files without exports can still be used for side effects, and be entry points of a package (or, tbh, entry points in a browser in an app).

**Comment by astorije** [comment]
I agree, which is why I think it would be nice to be able to configure those 2 things separately, since even the rule description explains the rule in 2 different parts: "Forbid modules without any export, and exports not imported by any modules"
